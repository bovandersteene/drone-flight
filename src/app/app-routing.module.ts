import { NgModule } from "@angular/core";
import { Routes, RouterModule } from "@angular/router";

const routes: Routes = [
  {
    path: "flight",
    loadChildren: "./flight/flight.module#FlightModule"
  },
  {
    path: "",
    redirectTo: "/flight",
    pathMatch: "full"
  }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule {}
