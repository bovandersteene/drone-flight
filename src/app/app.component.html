<mat-sidenav-container>
  <mat-sidenav #snav opened mode="side">
    <mat-nav-list>
      <a
        mat-list-item
        routerLinkActive="active"
        [routerLink]="['flight', flight.id]"
        *ngFor="let flight of (flights$ | async); trackBy: trackByFlight"
        >{{ flight.name }}</a
      >
    </mat-nav-list>
  </mat-sidenav>
  <mat-sidenav-content> <router-outlet></router-outlet> </mat-sidenav-content>
</mat-sidenav-container>

<a mat-fab routerLink="/flight" class="flight-new" color="primary">+</a>
<button mat-fab class="flight-info" (click)="openInfo()">i</button>
