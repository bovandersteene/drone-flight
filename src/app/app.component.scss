:host {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  .active {
    background-color: gray;
  }
  mat-sidenav {
    width: 200px;
  }
  mat-sidenav-container {
    height: 100%;
    mat-sidenav-content {
      padding: 20px;
    }
  }
  .flight-new {
    position: absolute;
    left: 20px;
    bottom: 20px;
    z-index: 500;
    font-size: 2.5em;
  }
  .flight-info {
    position: absolute;
    left: 100px;
    bottom: 20px;
    z-index: 500;
    font-size: 2.5em;
  }
}
