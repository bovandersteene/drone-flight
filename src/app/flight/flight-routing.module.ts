import { NgModule } from "@angular/core";
import { RouterModule, Routes } from "@angular/router";
import { FlightComponent } from "./flight/flight.component";

const routes: Routes = [
  {
    path: "",
    component: FlightComponent
  },
  {
    path: ":id",
    component: FlightComponent
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class FlightRoutingModule {}
