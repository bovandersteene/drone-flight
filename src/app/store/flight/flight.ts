import { LatLngLiteral } from "@agm/core";

export interface Flight {
  readonly id: number;
  readonly name: string;
  readonly location: LatLngLiteral;
  readonly path: Array<LatLngLiteral>;
}
