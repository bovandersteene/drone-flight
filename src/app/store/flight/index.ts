import { FEATURE_NAME } from "./flight.selector";
import { reducer } from "./flight.reducer";
import * as selectors from "./flight.selector";
export * from "./flight";

export const fromFlight = {
  FEATURE_NAME,
  reducer,
  selectors
};
