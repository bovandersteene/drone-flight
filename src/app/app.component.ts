import { Component, OnInit } from "@angular/core";
import { Flight } from "./store/flight";
import { FlightState } from "./store/flight/flight.state";
import { select, Store } from "@ngrx/store";
import { Observable } from "rxjs";
import { selectFlights } from "./store/flight/flight.selector";
import { MatDialog } from "@angular/material";
import { InfoComponent } from "./info/info.component";
import { LoadFlights } from "./store/flight/flight.actions";

@Component({
  selector: "app-root",
  templateUrl: "./app.component.html",
  styleUrls: ["./app.component.scss"]
})
export class AppComponent implements OnInit {
  flights$: Observable<Array<Flight>>;

  constructor(
    private readonly flightStore: Store<FlightState>,
    private readonly dialog: MatDialog
  ) {}

  ngOnInit(): void {
    this.flightStore.dispatch(new LoadFlights());
    this.flights$ = this.flightStore.pipe(select(selectFlights));
  }

  trackByFlight(flight: Flight): string {
    return flight.name;
  }

  openInfo() {
    this.dialog.open(InfoComponent);
  }
}
