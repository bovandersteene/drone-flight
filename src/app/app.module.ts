import { BrowserModule } from "@angular/platform-browser";
import { NgModule } from "@angular/core";

import { AppRoutingModule } from "./app-routing.module";
import { AppComponent } from "./app.component";
import { StoreModule } from "@ngrx/store";
import { metaReducers, reducers } from "./store/reducers";
import { EffectsModule } from "@ngrx/effects";
import { AppEffects } from "./store/app.effects";
import { BrowserAnimationsModule } from "@angular/platform-browser/animations";
import { CommonModule } from "@angular/common";
import { MatSidenavModule } from "@angular/material";
import { SharedModule } from "./shared/shared.module";
import { fromFlight } from "./store/flight";
import { StoreDevtoolsModule } from "@ngrx/store-devtools";
import { AgmCoreModule } from "@agm/core";
import { InfoComponent } from "./info/info.component";
import { FlightEffects } from "./store/flight/fligth.efffect";
import { environment } from "../environments/environment";

@NgModule({
  declarations: [AppComponent, InfoComponent],
  imports: [
    SharedModule,
    BrowserModule,
    AppRoutingModule,
    StoreModule.forRoot(reducers, { metaReducers }),
    StoreModule.forFeature(fromFlight.FEATURE_NAME, fromFlight.reducer),
    EffectsModule.forRoot([AppEffects]),
    EffectsModule.forFeature([FlightEffects]),
    StoreDevtoolsModule.instrument({
      maxAge: 5
    }),
    BrowserAnimationsModule,
    CommonModule,
    MatSidenavModule,
    AgmCoreModule.forRoot({
      // please get your own API key here:
      // https://developers.google.com/maps/documentation/javascript/get-api-key?hl=en
      apiKey: environment.mapKey
    })
  ],
  entryComponents: [InfoComponent],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule {}
