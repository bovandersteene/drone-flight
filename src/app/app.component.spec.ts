import { async, TestBed } from "@angular/core/testing";
import { RouterTestingModule } from "@angular/router/testing";
import { AppComponent } from "./app.component";
import { NO_ERRORS_SCHEMA } from "@angular/core";
import { MockStoreModule } from "../../test-helper";
import { Flight } from "./store/flight";
import { MatDialog } from "@angular/material";
import { LoadFlights } from "./store/flight/flight.actions";
import { Store } from "@ngrx/store";
import { FlightState } from "./store/flight/flight.state";

const flight = {
  id: 2,
  name: "Flight Lausanne"
} as Flight;

class MatDialogMock {
  open = jest.fn();
}

describe("AppComponent", () => {
  let matDialog: MatDialogMock;
  let store: Store<FlightState>;
  let app: AppComponent;
  let fixture;
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        MockStoreModule.forRoot("flight", {
          flights: [flight]
        })
      ],
      schemas: [NO_ERRORS_SCHEMA],
      declarations: [AppComponent],
      providers: [
        {
          provide: MatDialog,
          useClass: MatDialogMock
        }
      ]
    }).compileComponents();
  }));
  beforeEach(() => {
    fixture = TestBed.createComponent(AppComponent);
    matDialog = TestBed.get(MatDialog);
    store = TestBed.get(Store);
    app = fixture.debugElement.componentInstance;
  });
  it("should create the app", () => {
    expect(app).toBeTruthy();
  });

  it("should match the default snapshot", () => {
    fixture.detectChanges();
    expect(fixture.debugElement.nativeElement).toMatchSnapshot();
  });

  it("it should load the flights", () => {
    jest.spyOn(store, "dispatch");
    app.ngOnInit();
    expect(store.dispatch).toHaveBeenCalledWith(new LoadFlights());
  });

  describe("on info click", () => {
    it("should open the info dialog", () => {
      const compiled = fixture.debugElement.nativeElement;
      const infoButton = compiled.querySelector("button");
      infoButton.click();
      expect(matDialog.open).toHaveBeenCalled();
    });
  });
});
