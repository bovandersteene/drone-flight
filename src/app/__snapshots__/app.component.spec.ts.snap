// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should match the default snapshot 1`] = `
<div
  id="root1"
  ng-version="7.1.0"
>
  <mat-sidenav-container>
    <mat-sidenav
      mode="side"
      opened=""
    >
      <mat-nav-list>
        <!--bindings={
  "ng-reflect-ng-for-of": "[object Object]",
  "ng-reflect-ng-for-track-by": "function (flight) {\\n        /*"
}-->
        <a
          href="/flight/2"
          mat-list-item=""
          ng-reflect-router-link="flight,2"
          ng-reflect-router-link-active="active"
          routerlinkactive="active"
        >
          Flight Lausanne
        </a>
      </mat-nav-list>
    </mat-sidenav>
    <mat-sidenav-content>
      <router-outlet />
    </mat-sidenav-content>
  </mat-sidenav-container>
  <a
    class="flight-new"
    color="primary"
    href="/flight"
    mat-fab=""
    ng-reflect-router-link="/flight"
    routerlink="/flight"
  >
    +
  </a>
  <button
    class="flight-info"
    mat-fab=""
  >
    i
  </button>
</div>
`;
